hal/ia32/console.c

```diff
@@ -21,133 +21,343 @@
+				case '\b':
+				case '\177':
+					if (col) {
+						col--;
+					}
+					else if (row) {
+						row--;
+						col = halconsole_common.cols - 1;
+					}
+					break;
+
+				case '\n':
+					row++;
+				case '\r':
+					col = 0;
+					break;
+
+				case '\e':
+					hal_memset(halconsole_common.parms, 0, sizeof(halconsole_common.parms));
+					halconsole_common.parmi = 0;
+					halconsole_common.esc = esc_esc;
+					break;
 			}
-			break;
+		}
+		/* Process character according to escape sequence state */
+		else {
+			switch (halconsole_common.esc) {
+				case esc_init:
+					*(halconsole_common.vram + row * halconsole_common.cols + col) = (u16)halconsole_common.attr << 8 | c;
+					col++;
+					break;
+
+				case esc_esc:
+					switch (c) {
+						case '[':
+							hal_memset(halconsole_common.parms, 0, sizeof(halconsole_common.parms));
+							halconsole_common.parmi = 0;
+							halconsole_common.esc = esc_csi;
+							break;
+
+						default:
+							halconsole_common.esc = esc_init;
+							break;
+					}
+					break;
+
+				case esc_csi:
+					switch (c) {
+						case '0':
+						case '1':
+						case '2':
+						case '3':
+						case '4':
+						case '5':
+						case '6':
+						case '7':
+						case '8':
+						case '9':
+							halconsole_common.parms[halconsole_common.parmi] *= 10;
+							halconsole_common.parms[halconsole_common.parmi] += c - '0';
+							break;
+
+						case ';':
+							if (halconsole_common.parmi + 1 < sizeof(halconsole_common.parms))
+								halconsole_common.parmi++;
+							break;
+
+						case '?':
+							halconsole_common.esc = esc_csiqm;
+							break;
+
+						case 'H':
+							if (halconsole_common.parms[0] < 1)
+								halconsole_common.parms[0] = 1;
+							else if (halconsole_common.parms[0] > halconsole_common.rows)
+								halconsole_common.parms[0] = halconsole_common.rows;
+
+							if (halconsole_common.parms[1] < 1)
+								halconsole_common.parms[1] = 1;
+							else if (halconsole_common.parms[1] > halconsole_common.cols)
+								halconsole_common.parms[1] = halconsole_common.cols;
 
-		default:
-			*(console.vram + console.row * console.maxcol + console.col) = *p | attr << 8;
-			console.col++;
-			break;
+							row = halconsole_common.parms[0] - 1;
+							col = halconsole_common.parms[1] - 1;
+							halconsole_common.esc = esc_init;
+							break;
+
+						case 'J':
+							switch (halconsole_common.parms[0]) {
+								case 0:
+									console_memset(halconsole_common.vram + row * halconsole_common.cols + col, (u16)halconsole_common.attr << 8 | ' ', halconsole_common.cols * (halconsole_common.rows - row) - col);
+									break;
+
+								case 1:
+									console_memset(halconsole_common.vram, (u16)halconsole_common.attr << 8 | ' ', row * halconsole_common.cols + col + 1);
+									break;
+
+								case 2:
+									console_memset(halconsole_common.vram, (u16)halconsole_common.attr << 8 | ' ', halconsole_common.rows * halconsole_common.cols);
+									break;
+							}
```

Comment:
A line that is 250 characters long?
Maybe someting like that.
```suggestion
							; /* This empty statement must be */
							volatile u16 *vram = halconsole_common.vram;
							const u16 val = (u16)halconsole_common.attr << 8 | ' ';
							unsigned int n = 0;

							switch (halconsole_common.parms[0]) {
								case 0:
									vram += row * halconsole_common.cols + col;
									n = halconsole_common.cols * (halconsole_common.rows - row) - col;
									break;

								case 1:
									// vram = halconsole_common.vram;
									n = row * halconsole_common.cols + col + 1;
									break;

								case 2:
									// vram = halconsole_common.vram;
									n = halconsole_common.rows * halconsole_common.cols;
									break;
							}
							if (n > 0)
								console_memset(vram, val, n);

```

Reply:
This code is derived from [plo](https://github.com/phoenix-rtos/plo/blob/f5401788f72d58658a2aa1879ccfabc475cc900b/hal/ia32/console-vga.c#L79-L314) so it slightly differs from master branch.

`volatile` is correct, the vga hardware uses simultaneously from `halconsole_common.vram`.

We prefer not to write variables declaration inside `switch ... case`. I leave it as it is.

Reply:
But it seems to me that the hardware only reads from vram, it doesn't change its contents. You're just changing the contents of that memory.

Reply:
If hardware only reads this memory it must be volatile anyway - we need to make sure that we actually write to this memory

Reply:
If that's the case, what's changed that isn't in the master branch but is here.